### PR TITLE
Apply small optimizations to HTTPPropagator.extract

### PR DESCRIPTION
**Pull request by brettlangdon**
I was modifying this code to try and make the error message that we log better and realized that we could get also some performance wins by refactoring this code a little bit.

The main improvement I saw was that we are normalizing header keys 4 times per-extraction, which especially for `environ` we would potentially be iterating over and lowercasing a bunch ofkeys.

The other change was to remove the abstractions/method calls. (IMHO) they weren't really add much to the function, especially since we probably don't want those methods public anyways, and we never even tested them in isolation (we always test HTTPPropagator.extract).

The gains here aren't anything spectacular, this is definitely a micro-optimization. There are some users who might see a very tiny gain from this, since this code will run at the start of every web request we instrument.

### Before
```
------------------------------------------------------------------------------------------------ benchmark 'HTTPPropagator.extract': 5 tests -------------------------------------------------------------------------------------------------
Name (time in ns)                                Min                       Max                    Mean                  StdDev                  Median                     IQR            Outliers  OPS (Kops/s)            Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_extract_invalid_values                 737.4999 (1.0)        365,012.5000 (11.77)      7,851.5533 (6.73)       4,433.9763 (9.19)       6,949.9999 (6.69)         187.5001 (14.42)   1075;2018      127.3633 (0.15)      15129           8
test_extract_no_headers                     986.0000 (1.34)        31,020.0000 (1.0)        1,166.1404 (1.0)          482.3509 (1.0)        1,039.0000 (1.0)           13.0000 (1.0)      727;3379      857.5297 (1.0)        9488         100
test_extract_ideal_scenario               2,900.0003 (3.93)       209,150.0000 (6.74)      15,693.4945 (13.46)      7,680.8077 (15.92)     14,149.9995 (13.62)        350.0008 (26.92)    781;1749       63.7207 (0.07)      17954           2
test_extract_camel_dash_cased_large     611,999.9998 (829.83)   3,736,799.9994 (120.46)   697,848.2540 (598.43)   107,169.1775 (222.18)   685,199.9988 (659.48)   101,474.9982 (>1000.0)     71;39        1.4330 (0.00)       1575           1
test_extract_wsgi_large                 623,300.0004 (845.15)   2,215,500.0006 (71.42)    717,764.2400 (615.50)   106,179.3587 (220.13)   697,800.0001 (671.61)    93,900.0001 (>1000.0)     72;53        1.3932 (0.00)       1250           1
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```
### After

```
----------------------------------------------------------------------------------- benchmark 'HTTPPropagator.extract': 5 tests ------------------------------------------------------------------------------------
Name (time in us)                            Min                   Max                Mean             StdDev              Median                IQR            Outliers  OPS (Kops/s)            Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_extract_no_headers                   1.0690 (1.0)         29.6870 (1.0)        1.2528 (1.0)       0.4887 (1.0)        1.1240 (1.0)       0.0310 (1.0)      677;1445      798.1818 (1.0)        7850         100
test_extract_ideal_scenario               4.5833 (4.29)       474.4000 (15.98)      9.3008 (7.42)      5.5730 (11.40)      8.2000 (7.30)      0.2167 (6.99)    1081;1997      107.5176 (0.13)      16584           6
test_extract_invalid_values               8.1800 (7.65)       560.3800 (18.88)     11.7023 (9.34)      8.2343 (16.85)     10.0600 (8.95)      0.3000 (9.68)     539;1275       85.4535 (0.11)       8191           5
test_extract_camel_dash_cased_large     161.2000 (150.80)   3,020.7000 (101.75)   189.9914 (151.65)   61.6516 (126.16)   168.5000 (149.91)   19.7000 (635.48)   585;1071        5.2634 (0.01)       4817           1
test_extract_wsgi_large                 167.1000 (156.31)     654.4000 (22.04)    192.2771 (153.47)   42.1661 (86.28)    174.0000 (154.80)   14.2000 (458.06)    470;755        5.2008 (0.01)       3558           1
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

### CODE REVIEW

**Review by P403n1x87 — DISMISSED** [review]
💪  good stuff!

**Comment by brettlangdon** [comment]
I didn't do a release note on this, is one needed?

File: ddtrace/propagation/http.py [hunk 1]
```diff
@@ -123,28 +92,48 @@ def my_controller(url, headers):
...
+                    sampling_priority = int(sampling_priority)
+
+                return Context(
+                    trace_id=int(trace_id),
+                    span_id=int(parent_span_id),
+                    sampling_priority=sampling_priority,
+                    dd_origin=origin,
+                )
+            # If headers are invalid and cannot be parsed, return a new context and log the issue.
+            except (TypeError, ValueError):
```
Comment (Kyle-Verhoog) [inline ddtrace/propagation/http.py:128]: Can we add some test cases that covering these errors? I noticed that we didn't have any before so this code path is probably not covered: https://codecov.io/gh/DataDog/dd-trace-py/src/master/ddtrace/propagation/http.py

File: ddtrace/propagation/http.py [hunk 2]
```diff
@@ -56,52 +58,19 @@ def parent_call():
...
             try:
-                return normalized_headers[header]
+                return headers[header]
             except KeyError:
                 pass
 
         return default
 
-    @staticmethod
-    def extract_trace_id(headers):
```
Comment (Kyle-Verhoog) [inline ddtrace/propagation/http.py:70]: Glad to see these go!

**Review by Kyle-Verhoog — COMMENTED** [review]
Looks good! Just need a few test cases which already didn't exist. As for the changelog I think we can do without.

**Review by majorgreys — APPROVED** [review]

File: ddtrace/propagation/http.py [hunk 3]
```diff
@@ -123,28 +92,49 @@ def my_controller(url, headers):
...
+                return Context(
+                    # DEV: Do not allow `0` for trace id or span id, use None instead
+                    trace_id=int(trace_id) or None,
+                    span_id=int(parent_span_id) or None,
+                    sampling_priority=sampling_priority,
+                    dd_origin=origin,
+                )
+            # If headers are invalid and cannot be parsed, return a new context and log the issue.
+            except (TypeError, ValueError):
+                log.debug(
```
Comment (Kyle-Verhoog) [inline ddtrace/propagation/http.py:130]: It'd be good to have tests ensuring that debug logs are generated, but I think that's not required for the core change here. I'll make note of it in our backlog.

**Review by Kyle-Verhoog — APPROVED** [review]

**Pull request merged**